### PR TITLE
Added pucks calibration for upper arm/shoulders

### PR DESCRIPTION
**Pull request by druiz17**
Added pucks calibration for upper arm/shoulders

### CODE REVIEW

**Comment by druiz17** [comment]
Test plan-
- make sure you have pucks on feet, hips, and biceps
- in general, settings make sure toy have the configuration FeetHipsAndShoulder enabled, then preform the calibration step
- when performing the calibration try to be in the perfect T-pose and have the avatar hand pointing down to have the best results

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/149/HighFidelity-Beta-PR10542-8cb9bfcc81be9d4c3336f6dae25ee6814200e5a3.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/149/HighFidelity-Beta-PR10542-8cb9bfcc81be9d4c3336f6dae25ee6814200e5a3.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/186/HighFidelity-Beta-PR10542-cce19f702742e7f6f23b66f70bfa017f896a1526.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/186/HighFidelity-Beta-PR10542-cce19f702742e7f6f23b66f70bfa017f896a1526.dmg

**Comment by anshuman64** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/191/HighFidelity-Beta-PR10542-cce19f702742e7f6f23b66f70bfa017f896a1526.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/191/HighFidelity-Beta-PR10542-cce19f702742e7f6f23b66f70bfa017f896a1526.dmg

File: libraries/animation/src/Rig.cpp [hunk 1]
```diff
@@ -1047,6 +1047,23 @@ void Rig::updateFromHeadParameters(const HeadParameters& params, float dt) {
     } else {
         _animVars.set("spine2Type", (int)IKTarget::Type::Unknown);
     }
+
+    if (params.leftArmEnabled) {
+        _animVars.set("leftArmType", (int)IKTarget::Type::RotationAndPosition);
+        _animVars.set("leftArmPosition", params.leftArmPosition);
+        _animVars.set("leftArmRotation", params.leftArmRotation);
+
```
Comment (hyperlogic) [inline libraries/animation/src/Rig.cpp]: remove empty line

File: plugins/openvr/src/ViveControllerManager.cpp [hunk 2]
```diff
@@ -333,34 +340,21 @@ void ViveControllerManager::InputDevice::calibrate(const controller::InputCalibr
...
+        calibrateHips(defaultToReferenceMat, inputCalibration);
     } else if (_config == Config::FeetHipsAndChest && puckCount >= MIN_FEET_HIPS_CHEST) {
-        _jointToPuckMap[controller::HIPS] = _validTrackedObjects[HIP].first;
-        _pucksOffset[_validTrackedObjects[HIP].first] = computeOffset(defaultToReferenceMat, inputCalibration.defaultHips, _validTrackedObjects[HIP].second);
-        _jointToPuckMap[controller::SPINE2] = _validTrackedObjects[CHEST].first;
-        _pucksOffset[_validTrackedObjects[CHEST].first] = computeOffset(defaultToReferenceMat, inputCalibration.defaultSpine2, _validTrackedObjects[CHEST].second);
+        calibrateFeet(defaultToReferenceMat, inputCalibration);
+        calibrateHips(defaultToReferenceMat, inputCalibration);
+        calibrateChest(defaultToReferenceMat, inputCalibration);
+    } else if (_config == Config::FeetHipsAndShoulders && puckCount >= MIN_FEET_HIPS_SHOULDERS){
```
Comment (hyperlogic) [inline plugins/openvr/src/ViveControllerManager.cpp]: Add a space between closing paren and open curly brace.

File: plugins/openvr/src/ViveControllerManager.cpp [hunk 3]
```diff
@@ -615,6 +611,57 @@ void ViveControllerManager::InputDevice::hapticsHelper(float deltaTime, bool lef
...
+    auto& firstFoot = _validTrackedObjects[FIRST_FOOT];
+    auto& secondFoot = _validTrackedObjects[SECOND_FOOT];
+    controller::Pose& firstFootPose = firstFoot.second;
+    controller::Pose& secondFootPose = secondFoot.second;
+    
+    if (firstFootPose.translation.x < secondFootPose.translation.x) {
+        _jointToPuckMap[controller::LEFT_FOOT] = firstFoot.first;
+        _pucksOffset[firstFoot.first] = computeOffset(defaultToReferenceMat, inputCalibration.defaultLeftFoot, firstFootPose);
+        _jointToPuckMap[controller::RIGHT_FOOT] = secondFoot.first;
+            _pucksOffset[secondFoot.first] = computeOffset(defaultToReferenceMat, inputCalibration.defaultRightFoot, secondFootPose);
```
Comment (hyperlogic) [inline plugins/openvr/src/ViveControllerManager.cpp]: indenting is off here, tabs?

File: plugins/openvr/src/ViveControllerManager.cpp [hunk 4]
```diff
@@ -615,6 +611,57 @@ void ViveControllerManager::InputDevice::hapticsHelper(float deltaTime, bool lef
...
+}
+
+void ViveControllerManager::InputDevice::calibrateChest(glm::mat4& defaultToReferenceMat, const controller::InputCalibrationData& inputCalibration) {
+    _jointToPuckMap[controller::SPINE2] = _validTrackedObjects[CHEST].first;
+    _pucksOffset[_validTrackedObjects[CHEST].first] = computeOffset(defaultToReferenceMat, inputCalibration.defaultSpine2, _validTrackedObjects[CHEST].second);
+}
+
+void ViveControllerManager::InputDevice::calibrateShoulders(glm::mat4& defaultToReferenceMat, const controller::InputCalibrationData& inputCalibration,
+                                                            int firstShoulderIndex, int secondShoulderIndex) {
+    qDebug() << "calibrate shoulders";
```
Comment (hyperlogic) [inline plugins/openvr/src/ViveControllerManager.cpp]: remove debug print

File: plugins/openvr/src/ViveControllerManager.cpp [hunk 5]
```diff
@@ -658,7 +707,7 @@ void ViveControllerManager::InputDevice::createPreferences() {
         auto getter = [this]()->QString { return _configStringMap[_preferedConfig]; };
         auto setter = [this](const QString& value) { setConfigFromString(value); saveSettings(); };
         auto preference = new ComboBoxPreference(VIVE_PUCKS_CONFIG, "Configuration", getter, setter);
-        QStringList list = (QStringList() << "Auto" << "Feet" << "FeetAndHips" << "FeetHipsAndChest");
+        QStringList list = (QStringList() << "Auto" << "Feet" << "FeetAndHips" << "FeetHipsAndChest" << "FeetHipsAndShoulders");
```
Comment (hyperlogic) [inline plugins/openvr/src/ViveControllerManager.cpp]: I could be wrong, but QStringList supports static initializer lists so you could do something like this:

    QStringList list = { "Auto", "Feet", "FeetAndHips", "FeetHipsAndChest", "FeetHipsAndShoulders" };

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/198/HighFidelity-Beta-PR10542-d317d9236fd248033d75965a94cc50b56ef9e102.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/198/HighFidelity-Beta-PR10542-d317d9236fd248033d75965a94cc50b56ef9e102.dmg

**Pull request merged**